File: internal/helpers/password_test.go
// This file is part of the happyDomain (R) project.
// Copyright (c) 2020-2025 happyDomain
// Authors: Pierre-Olivier Mercier, et al.
//
// This program is offered under a commercial and under the AGPL license.
// For commercial licensing, contact us at <contact@happydomain.org>.
//
// For AGPL licensing:
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

package helpers

import (
	"strings"
	"testing"
)

func TestGeneratePassword(t *testing.T) {
	t.Run("generates 12 character password", func(t *testing.T) {
		password, err := GeneratePassword()
		if err != nil {
			t.Fatalf("GeneratePassword() returned error: %v", err)
		}

		if len(password) != 12 {
			t.Errorf("GeneratePassword() generated password of length %d, want 12", len(password))
		}
	})

	t.Run("password does not contain replaced characters", func(t *testing.T) {
		forbiddenChars := []string{"v", "u", "l", "1", "o", "O", "0", "/"}

		for range 100 {
			password, err := GeneratePassword()
			if err != nil {
				t.Fatalf("GeneratePassword() returned error: %v", err)
			}

			for _, char := range forbiddenChars {
				if strings.Contains(password, char) {
					t.Errorf("GeneratePassword() generated password containing forbidden character %q: %s", char, password)
				}
			}
		}
	})

	t.Run("password contains replacement characters", func(t *testing.T) {
		replacementChars := []string{"*", "(", "%", "?", "@", "!", ">", "^"}
		foundChars := make(map[string]bool)

		for range 1000 {
			password, err := GeneratePassword()
			if err != nil {
				t.Fatalf("GeneratePassword() returned error: %v", err)
			}

			for _, char := range replacementChars {
				if strings.Contains(password, char) {
					foundChars[char] = true
				}
			}
		}

		if len(foundChars) == 0 {
			t.Error("GeneratePassword() did not use any replacement characters in 1000 attempts")
		}
	})

	t.Run("generates different passwords", func(t *testing.T) {
		passwords := make(map[string]bool)
		iterations := 100

		for range iterations {
			password, err := GeneratePassword()
			if err != nil {
				t.Fatalf("GeneratePassword() returned error: %v", err)
			}
			passwords[password] = true
		}

		if len(passwords) < iterations {
			duplicates := iterations - len(passwords)
			t.Errorf("GeneratePassword() generated %d duplicate passwords out of %d", duplicates, iterations)
		}
	})

	t.Run("password uses valid characters", func(t *testing.T) {
		validChars := "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+*(%?@!>^="

		for range 100 {
			password, err := GeneratePassword()
			if err != nil {
				t.Fatalf("GeneratePassword() returned error: %v", err)
			}

			for _, r := range password {
				if !strings.ContainsRune(validChars, r) {
					t.Errorf("GeneratePassword() generated password with invalid character %q: %s", r, password)
				}
			}
		}
	})

	t.Run("password has good entropy", func(t *testing.T) {
		charCounts := make(map[rune]int)
		iterations := 1000

		for range iterations {
			password, err := GeneratePassword()
			if err != nil {
				t.Fatalf("GeneratePassword() returned error: %v", err)
			}

			for _, r := range password {
				charCounts[r]++
			}
		}

		uniqueChars := len(charCounts)
		if uniqueChars < 30 {
			t.Errorf("GeneratePassword() used only %d unique characters in %d passwords, expected more variety", uniqueChars, iterations)
		}
	})

	t.Run("password ends with valid character", func(t *testing.T) {
		for range 100 {
			password, err := GeneratePassword()
			if err != nil {
				t.Fatalf("GeneratePassword() returned error: %v", err)
			}

			lastChar := password[len(password)-1]
			if lastChar == '=' {
				t.Log("Password contains base64 padding character '=' at the end (acceptable)")
			}
		}
	})
}

func TestGeneratePasswordNonEmpty(t *testing.T) {
	password, err := GeneratePassword()
	if err != nil {
		t.Fatalf("GeneratePassword() returned error: %v", err)
	}

	if password == "" {
		t.Error("GeneratePassword() returned empty password")
	}
}

func TestGeneratePasswordConsistentLength(t *testing.T) {
	lengths := make(map[int]int)

	for range 1000 {
		password, err := GeneratePassword()
		if err != nil {
			t.Fatalf("GeneratePassword() returned error: %v", err)
		}
		lengths[len(password)]++
	}

	if len(lengths) != 1 {
		t.Errorf("GeneratePassword() generated passwords with varying lengths: %v", lengths)
	}

	for length := range lengths {
		if length != 12 {
			t.Errorf("GeneratePassword() generated password of length %d, want 12", length)
		}
	}
}

func TestGeneratePasswordReplacements(t *testing.T) {
	replacements := map[string]string{
		"v": "*",
		"u": "(",
		"l": "%",
		"1": "?",
		"o": "@",
		"O": "!",
		"0": ">",
		"/": "^",
	}

	t.Run("verifies character replacements", func(t *testing.T) {
		for range 100 {
			password, err := GeneratePassword()
			if err != nil {
				t.Fatalf("GeneratePassword() returned error: %v", err)
			}

			for original := range replacements {
				if strings.Contains(password, original) {
					t.Errorf("Password contains unreplaced character %q: %s", original, password)
				}
			}
		}
	})
}

func BenchmarkGeneratePassword(b *testing.B) {
	for b.Loop() {
		_, err := GeneratePassword()
		if err != nil {
			b.Fatalf("GeneratePassword() returned error: %v", err)
		}
	}
}

func BenchmarkGeneratePasswordParallel(b *testing.B) {
	b.RunParallel(func(pb *testing.PB) {
		for pb.Next() {
			_, err := GeneratePassword()
			if err != nil {
				b.Fatalf("GeneratePassword() returned error: %v", err)
			}
		}
	})
}

File: web/src/lib/password.test.ts
// This file is part of the happyDomain (R) project.
// Copyright (c) 2022-2026 happyDomain
//
// This program is offered under a commercial and under the AGPL license.
// For commercial licensing, contact us at <contact@happydomain.org>.
//
// For AGPL licensing:
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

import { describe, it, expect } from "vitest";
import { checkWeakPassword, checkPasswordConfirmation } from "./password";

describe("checkWeakPassword", () => {
    it("returns undefined for an empty password", () => {
        expect(checkWeakPassword("")).toBeUndefined();
    });

    it("returns true for a strong password with special char", () => {
        expect(checkWeakPassword("Abcdefg1!")).toBe(true);
    });

    it("returns true for a strong password without special char but long enough", () => {
        expect(checkWeakPassword("Abcdefghij1")).toBe(true);
    });

    it("returns false when shorter than 11 chars and missing a special char", () => {
        expect(checkWeakPassword("Abcdefg1")).toBe(false);
    });

    it("returns false when shorter than 8 chars", () => {
        expect(checkWeakPassword("Ab1!")).toBe(false);
    });

    it("returns false when missing an uppercase letter", () => {
        expect(checkWeakPassword("abcdefg1!")).toBe(false);
    });

    it("returns false when missing a lowercase letter", () => {
        expect(checkWeakPassword("ABCDEFG1!")).toBe(false);
    });

    it("returns false when missing a digit", () => {
        expect(checkWeakPassword("Abcdefgh!")).toBe(false);
    });

    it("returns false for only letters, regardless of length", () => {
        expect(checkWeakPassword("Abcdefghijklmno")).toBe(false);
    });

    it("returns false for only digits", () => {
        expect(checkWeakPassword("12345678901")).toBe(false);
    });

    it("accepts exactly 8 characters with a special char", () => {
        expect(checkWeakPassword("Abcdefg1!")).toBe(true);
    });

    it("rejects 7 characters even with all classes and a special char", () => {
        expect(checkWeakPassword("Abcde1!")).toBe(false);
    });

    it("accepts exactly 11 characters without a special char", () => {
        expect(checkWeakPassword("Abcdefghij1")).toBe(true);
    });

    it("rejects 10 characters without a special char", () => {
        expect(checkWeakPassword("Abcdefghi1")).toBe(false);
    });

    it("treats underscore as a word character, not special", () => {
        // \W does not match underscore, so length must reach 11 to pass
        expect(checkWeakPassword("Abcdefg_1")).toBe(false);
        expect(checkWeakPassword("Abcdefghi_1")).toBe(true);
    });

    it("accepts unicode/space as a special character", () => {
        expect(checkWeakPassword("Abcdefg 1")).toBe(true);
    });

    it("handles a whitespace-only password as non-empty", () => {
        expect(checkWeakPassword("        ")).toBe(false);
    });
});

describe("checkPasswordConfirmation", () => {
    it("returns undefined when confirmation is empty", () => {
        expect(checkPasswordConfirmation("password", "")).toBeUndefined();
    });

    it("returns undefined when both are empty", () => {
        expect(checkPasswordConfirmation("", "")).toBeUndefined();
    });

    it("returns true when password and confirmation match", () => {
        expect(checkPasswordConfirmation("password", "password")).toBe(true);
    });

    it("returns false when password and confirmation differ", () => {
        expect(checkPasswordConfirmation("password", "differen")).toBe(false);
    });

    it("returns false when confirmation is non-empty but password is empty", () => {
        expect(checkPasswordConfirmation("", "password")).toBe(false);
    });

    it("is case-sensitive", () => {
        expect(checkPasswordConfirmation("Password", "password")).toBe(false);
    });
});
